src/Misc/khronosTextureContainer2.ts

```diff
@@ -1,69 +1,227 @@
+                        const mipmap = data.mipmaps[mip];
+                        if (mipmap) {
+                            buffers.push(mipmap.data.buffer);
+                        }
+                    }
+                    this._createTexture(data, internalTexture);
+                    resolve();
+                }).catch((reason: any) => {
+                    reject({ message: reason });
+                });
+            } catch (err) {
+                reject({ message: err });
             }
         });
     }
 
-    private _determineTranscodeFormat(transcodeTarget: any, caps: EngineCapabilities): number {
-        if (caps.s3tc) {
-            return transcodeTarget.BC1_OR_3;
+    /**
+     * Stop all async operations and release resources.
+     */
+    public dispose(): void {
+        if (KhronosTextureContainer2._WorkerPoolPromise) {
+            KhronosTextureContainer2._WorkerPoolPromise.then((workerPool) => {
+                workerPool.dispose();
+            });
+        }
+
+        delete KhronosTextureContainer2._WorkerPoolPromise;
+    }
+
+    protected _createTexture(data: any /* IEncodedData */, internalTexture: InternalTexture) {
+        this._engine._bindTextureDirectly(this._engine._gl.TEXTURE_2D, internalTexture);
+
+        if (data.transcodedFormat === 0x8058 /* RGBA8 */) {
+            internalTexture.type = Constants.TEXTURETYPE_UNSIGNED_BYTE;
+            internalTexture.format = Constants.TEXTUREFORMAT_RGBA;
+        } else {
+            internalTexture.format = data.transcodedFormat;
         }
-        else if (caps.etc2) {
-            return transcodeTarget.ETC;
+
+        internalTexture._gammaSpace = data.isInGammaSpace;
+
+        if (data.errors) {
+            throw new Error("KTX2 container - could not transcode the data. " + data.errors);
         }
 
-        throw new Error("No compatible format available");
+        for (let t = 0; t < data.mipmaps.length; ++t) {
+            let mipmap = data.mipmaps[t];
+
+            if (!mipmap || !mipmap.data) {
+                throw new Error("KTX2 container - could not transcode one of the image");
+            }
+
+            if (data.transcodedFormat === 0x8058 /* RGBA8 */) {
+                // uncompressed RGBA
+                internalTexture.width = mipmap.width; // need to set width/height so that the call to _uploadDataToTextureDirectly uses the right dimensions
+                internalTexture.height = mipmap.height;
+
+                this._engine._uploadDataToTextureDirectly(internalTexture, mipmap.data, 0, t, undefined, true);
+            } else {
+                this._engine._uploadCompressedDataToTextureDirectly(internalTexture, data.transcodedFormat, mipmap.width, mipmap.height, mipmap.data, 0, t);
+            }
+        }
+
+        internalTexture.width = data.mipmaps[0].width;
+        internalTexture.height = data.mipmaps[0].height;
+        internalTexture.generateMipMaps = data.mipmaps.length > 1;

```

Comment:
This logic is backwards? If there are mipmaps, you don't need to generate them right?

Reply:
The thing is that it is misleading because I have to set `generateMipMaps = true` for the code in Babylon to work!

`generateMipMaps = true` **won't** make the engine generate mipmaps when the texture is compressed:
```
if (!noMipmap && !isCompressed) {
    gl.generateMipmap(gl.TEXTURE_2D);
}
```
This is the code in `_prepareWebGLTextureContinuation`.

I need `generateMipMaps` to be true so that the right `TEXTURE_MIN_FILTER` value is retrieved when calling `ThinEngine._getSamplingParameters(samplingMode: number, generateMipMaps: boolean)`

Reply:
Oh right, I vaguely remember this being weird. Maybe file an issue to fix this after this PR is merged?

Reply:
Done in #8863